testbed/tests/widgets/test_scrollcontainer.py

```diff
@@ -0,0 +1,453 @@
+from unittest.mock import Mock
+
+import pytest
+from pytest import approx
+
+import toga
+from toga.colors import CORNFLOWERBLUE, REBECCAPURPLE, TRANSPARENT
+from toga.style.pack import COLUMN, ROW, Pack
+
+from .properties import (  # noqa: F401
+    test_background_color,
+    test_background_color_reset,
+    test_background_color_transparent,
+    test_enable_noop,
+    test_flex_widget_size,
+    test_focus_noop,
+)
+
+# DPI note: document_width and document_height may be subject to cumulative rounding
+# error. This should be fixed by #2020, at which time the `approx` calls with rel=0.01
+# can be removed. The calls with abs=1 will probably have to stay.
```

Comment:
Before adding the `approx` calls, the Android tests were passing on a device with a scale factor of 2, but failed on Briefcase's default emulator, which has a scale factor of 2.625:

```
I/python.stdout: =================================== FAILURES ===================================
I/python.stdout: ______________________________ test_clear_content ______________________________
I/python.stdout: Traceback (most recent call last):
I/python.stdout:   File "/data/data/org.beeware.toga.testbed/files/chaquopy/AssetFinder/app/tests/widgets/test_scrollcontainer.py", line 77, in test_clear_content
I/python.stdout:     assert probe.document_height == 6000, f"FIXME {probe.document_height} != {6000}"
I/python.stdout: AssertionError: FIXME 5942.857142857143 != 6000
I/python.stdout: _____________________________ test_vertical_scroll _____________________________
I/python.stdout: Traceback (most recent call last):
I/python.stdout:   File "/data/data/org.beeware.toga.testbed/files/chaquopy/AssetFinder/app/tests/widgets/test_scrollcontainer.py", line 231, in test_vertical_scroll
I/python.stdout:     assert probe.document_height == 6000, f"FIXME {probe.document_height} != {6000}"
I/python.stdout: AssertionError: FIXME 5942.857142857143 != 6000
I/python.stdout: ____________________________ test_horizontal_scroll ____________________________
I/python.stdout: Traceback (most recent call last):
I/python.stdout:   File "/data/data/org.beeware.toga.testbed/files/chaquopy/AssetFinder/app/tests/widgets/test_scrollcontainer.py", line 297, in test_horizontal_scroll
I/python.stdout:     assert probe.document_width == 20000, f"FIXME {probe.document_width} != {20000}"
I/python.stdout: AssertionError: FIXME 19961.904761904763 != 20000
I/python.stdout: _______________________________ test_scroll_both _______________________________
I/python.stdout: Traceback (most recent call last):
I/python.stdout:   File "/data/data/org.beeware.toga.testbed/files/chaquopy/AssetFinder/app/tests/widgets/test_scrollcontainer.py", line 375, in test_scroll_both
I/python.stdout:     assert probe.document_width == 2040, f"FIXME {probe.document_width} != {2040}"
I/python.stdout: AssertionError: FIXME 2039.6190476190477 != 2040
```

Notice that the test where the document size is determined by just 1 widget is off by about 0.5 pixels, while the tests with 100 widgets side by side are off by about 50 pixels.

Reply:
This should be improved by #2020.